pkg/lib/catalogsource/catalogsource_update.go

```diff
@@ -0,0 +1,173 @@
+	if err != nil {
+		logger.WithError(err).Error("UpdateStatus - error getting latest CatalogSource... cannot update image reference")
+		return err
+	}
+
+	// make copy (even though we're not making changes and using the status values as-is)
+	out := latest.DeepCopy()
+	out.Status = catsrc.Status
+
+	// make the status update if possible
+	if _, err := client.OperatorsV1alpha1().CatalogSources(out.GetNamespace()).UpdateStatus(context.TODO(), out, metav1.UpdateOptions{}); err != nil {
+		logger.WithError(err).Error("UpdateStatus - error while setting CatalogSource status")
+		return err
+	}
+
+	return nil
+}
+
+/* UpdateStatusCondition can be used to safely update the status conditions for the provided catalog source.
+This function will make no other changes to the status.
+
+• logger: used to log errors only
+
+• client: used to fetch / update catalog source status
+
+• catsrc: the CatalogSource to use as a source for status updates.
+
+• conditions: condition values to be updated
+*/
+func UpdateStatusCondition(logger *logrus.Entry, client versioned.Interface, catsrc *v1alpha1.CatalogSource, conditions ...metav1.Condition) error {
+	mu.Lock()
+	defer mu.Unlock()
+
+	// get the absolute latest update of this catalog source in case it changed
+	latest, err := client.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).Get(context.TODO(), catsrc.GetName(), metav1.GetOptions{})
+	if err != nil {
+		logger.WithError(err).Error("UpdateStatusCondition - error getting latest CatalogSource... cannot update image reference")
+		return err
+	}
+
+	// make copy and update the image and conditions only
+	out := latest.DeepCopy()
+
+	for _, condition := range conditions {
+		meta.SetStatusCondition(&out.Status.Conditions, condition)
+	}
+
+	// don't bother updating if no changes were made
+	if reflect.DeepEqual(out.Status.Conditions, latest.Status.Conditions) {
+		logger.Debug("UpdateStatusCondition - request to update status conditions did not result in any changes, so updates were not made")
+		return nil
+	}
+
+	// make the update if possible
+	if _, err := client.OperatorsV1alpha1().CatalogSources(catsrc.GetNamespace()).UpdateStatus(context.TODO(), out, metav1.UpdateOptions{}); err != nil {
+		logger.WithError(err).Error("UpdateStatusCondition - unable to update CatalogSource image reference")
+		return err
+	}
+	return nil
+}
+
+/* UpdateImageReferenceAndStatusCondition can be used to safely update the image reference and status conditions for the provided catalog source.
+This function will make no other changes to the catalog source.
+
+• logger: used to log errors only
+
+• client: used to fetch / update catalog source
+
+• catsrc: the CatalogSource to use as a source for image and status condition updates.
+
+• conditions: condition values to be updated
+*/
+func UpdateImageReferenceAndStatusCondition(logger *logrus.Entry, client versioned.Interface, catsrc *v1alpha1.CatalogSource, conditions ...metav1.Condition) error {
```

Comment:
Its mostly about shared resource updating and locking... see https://github.com/operator-framework/operator-lifecycle-manager/pull/2281#discussion_r677511023

Reply:
Ok... so the `Update*` functions I added are almost all dealing with status updates only (i.e. `client.OperatorsV1alpha1().CatalogSources(...).UpdateStatus(...)`). The only one that updates anything in the `spec` is this new controller (to update the image reference) so it uses `client.OperatorsV1alpha1().CatalogSources(...).Update(...)`. 

I think you're saying that patch (i.e. `client.OperatorsV1alpha1().CatalogSources(...).Patch(...)`) is preferable over `client.OperatorsV1alpha1().CatalogSources(...).Update(...)`. 

It does not seem like the patch function should be used for status updates,  might make sense for the `UpdateImageReferenceAndStatusCondition` function since it updates the image reference. 

Honestly it sounds like you're saying I don't need to

Reply:
Resolving this conversation... can continue in other thread